### ISSUE TITLE
Error: duplicate entry, AnnotationVisitor.class

### ISSUE THREAD

**Issue by denebchorny** [jayway/JsonPath#224]
Hi Team,

I know this question was issued here: https://github.com/jayway/JsonPath/issues/213
but  I haven't found a solution yet.

I would appreciate your support on this issue. I'll bet I'm not the only one with this.

My understanding is that I am trying to package the same class twice or there are duplicate references to the same API (Likely different version numbers), I am not sure at all.

At first time, I want to know what's the escential of this kind of problems " java.util.zip.ZipException".
And second, I looking for help on how to use this library.

@kallestenflo said: "This is a json-smart issue https://github.com/netplex/json-smart-v2. Try configuring JsonPath to use an other JsonProvider. This must be done before you access JsonPath the first time."

How can I do this?, developing for android I am an amateur. I undertand a lot of things but in this case my mind is blowing.

Please, tell me there is a solution for this. T.T

Thanks in advance

**Comment by kallestenflo** [comment]
I don't have any Android experience so I'm not sure what is wrong. 

To use an other JsonProvider you can  include the code below when you initialize your application. It should only run once. You also need to exclude the `net.minidev:json-smart:2.2.1` dependency from your build and include `com.fasterxml.jackson.core:jackson-databind:2.6.3`.

```
Configuration.setDefaults(new Configuration.Defaults() {

    private final JsonProvider jsonProvider = new JacksonJsonProvider();
    private final MappingProvider mappingProvider = new JacksonMappingProvider();

    @Override
    public JsonProvider jsonProvider() {
        return jsonProvider;
    }

    @Override
    public MappingProvider mappingProvider() {
        return mappingProvider;
    }

    @Override
    public Set<Option> options() {
        return EnumSet.noneOf(Option.class);
    }
});
```

**Comment by denebchorny** [comment]
Perfect Mr. @kallestenflo,

Two things more.

1) When you say: "You also need to exclude the net.minidev:json-smart:2.2.1 dependency from your build and include com.fasterxml.jackson.core:jackson-databind:2.6.3"

How can I exclude and Include them?.

This is my Build.Gradle (app):

```xml
repositories {
    jcenter()
    mavenCentral()
    maven { url "https://jitpack.io" }
    maven { url "http://jzaccone.github.io/SlidingMenu-aar" }
    maven { url "https://s3.amazonaws.com/repo.commonsware.com" }
    maven { url "http://files.couchbase.com/maven2/" }
}

android {
    compileSdkVersion 23
    buildToolsVersion "23.0.2"
    defaultConfig {
        applicationId "com.xpectra.handheld"
        minSdkVersion 15
        targetSdkVersion 23
        versionCode 1
        versionName "1.0"
        multiDexEnabled true
    }
    dexOptions {
        incremental true
        javaMaxHeapSize "4g"
    }
    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
        }
    }
    packagingOptions {
        exclude 'META-INF/ASL2.0'
        exclude 'META-INF/LICENSE'
        exclude 'META-INF/NOTICE'
    }
}

dependencies {
    compile fileTree(dir: 'libs', include: ['*.jar'])
    compile 'com.android.support:design:23.1.1'
    compile 'com.android.support:cardview-v7:23.1.1'
    compile 'com.android.support:recyclerview-v7:23.1.1'
    compile 'com.jakewharton:butterknife:7.0.1'
    compile 'uk.co.chrisjenx:calligraphy:2.1.0'
    compile 'com.github.bumptech.glide:glide:3.6.1'
    compile 'com.squareup.retrofit2:retrofit:2.0.0'
    compile 'com.squareup.retrofit2:converter-gson:2.0.0'
    compile 'com.google.code.gson:gson:2.6.2'
    compile 'com.google.android.gms:play-services-maps:8.1.0'
    compile 'com.google.android.gms:play-services-location:8.1.0'
    compile 'joda-time:joda-time:2.9.1'
    compile 'de.greenrobot:eventbus:2.4.0'
    compile 'com.couchbase.lite:couchbase-lite-android:1.2.1'

   // JSON_PATH	
   compile 'com.jayway.jsonpath:json-path:2.2.0'
}
```

I am not including any provider. Just GSON because I'm using it in retrofit.
I am importing this library via Gradle. 

2) If I want to use GSON as provider I just need to define it in the configuration you gave me before, is it? 

Excuse me if I ask you nonsense or dumb questions but I haven't faced with something similar.

**Comment by kallestenflo** [comment]
Check this out https://docs.gradle.org/current/userguide/dependency_management.html (23.4.7)

**Comment by dadoonet** [comment]
For the record, to avoid JAR Hell issues (duplicate classnames), I had to add some exclusions to my maven project file:

```xml
        <dependency>
            <groupId>com.jayway.jsonpath</groupId>
            <artifactId>json-path</artifactId>
            <version>2.2.0</version>
            <exclusions>
                <exclusion>
                    <groupId>org.ow2.asm</groupId>
                    <artifactId>asm</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
```

Hope this helps others.

**Comment by zishanj** [comment]
@dadoonet this helps solved my issue too...

**Comment by zishanj** [comment]
In gradle I have excluded it like this

`compile ('com.jayway.jsonpath:json-path:2.2.0') {
        exclude group: 'org.ow2.asm'
    }`

**Comment by jochenberger** [comment]
@kallestenflo this was fixed via #315

**Comment by kallestenflo** [comment]
Fixed via #315
